server/src/main/java/org/elasticsearch/index/query/first/FirstQuery.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.query.first;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.Matches;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.ScorerSupplier;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.Scorer;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Set;
+
+
+public class FirstQuery extends Query {
+    private final Query[] queries;
+
+    public FirstQuery(Query[] queries) {
+        this.queries = queries;
+    }
+
+    public Query[] getQueries() {
+        return queries;
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        if (queries.length == 1) {
+            return queries[0];
+        }
+        Query[] rewrittenQueries = new Query[queries.length];
+        boolean rewritten = false;
+        for (int i = 0; i < queries.length; i++) {
+            Query rewrittenQuery = queries[i].rewrite(reader);
+            rewrittenQueries[i] = rewrittenQuery;
+            if (rewrittenQuery != queries[i]) {
+                rewritten = true;
+            }
+        }
+        if (rewritten) {
+            return new FirstQuery(rewrittenQueries);
+        }
+        return this;
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+        QueryVisitor v = visitor.getSubVisitor(BooleanClause.Occur.SHOULD, this);
```

Comment:
Also just out of curiosity: what's the semantics of using `SHOULD` here? I see e.g. also `BlendedTermQuery` uses this, but others like ScriptScoreQuery use `MUST`. Just briefly scanned some QueryVisitor implemenations and wasn't able to figure out the difference, if there is any.

Reply:
I am not 100% sure here, @romseygeek  would the best to confirm.  

I think `visit(QueryVisitor visitor)` is primarily used for highlighting, where there is no difference between `SHOULD` and `MUST`.  But I also see it in the [TopDocsCollectorContext](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorContext.java#L474] to check if the query tree contains any query with infinite max score, and here we are checking for `MUST` condition.

I guess, I put `SHOULD` as if a doc matches the 1st query, all other queries are optional.

Reply:
Percolator uses `visit` as well to build its query index.  I think in this case we can view this query as a specialised disjunction; a match on any of its clauses will result in a match overall.